core/consensus/pow/src/lib.rs

```diff
@@ -0,0 +1,399 @@
+	) -> Result<(BlockImportParams<B>, Option<Vec<(CacheKeyId, Vec<u8>)>>), String> {
+		let inherent_data = self.inherent_data_providers
+			.create_inherent_data().map_err(String::from)?;
+		let timestamp_now = inherent_data.timestamp_inherent_data().map_err(String::from)?;
+
+		let best_hash = self.client.info().best_hash;
+		let hash = header.hash();
+		let parent_hash = *header.parent_hash();
+		let best_aux = PowAux::read(self.client.as_ref(), &best_hash)?;
+		let mut aux = PowAux::read(self.client.as_ref(), &parent_hash)?;
+
+		let (checked_header, difficulty, seal) = self.check_header::<B>(
+			header,
+			BlockId::Hash(parent_hash),
+		)?;
+		aux.total_difficulty = aux.total_difficulty.saturating_add(difficulty);
+
+		if let Some(inner_body) = body.take() {
+			let block = B::new(checked_header.clone(), inner_body);
+
+			self.check_inherents(
+				block.clone(),
+				BlockId::Hash(parent_hash),
+				inherent_data,
+				timestamp_now
+			)?;
+
+			let (_, inner_body) = block.deconstruct();
+			body = Some(inner_body);
+		}
+		let key = aux_key(&hash);
+		let import_block = BlockImportParams {
+			origin,
+			header: checked_header,
+			post_digests: vec![seal],
+			body,
+			finalized: false,
+			justification,
+			auxiliary: vec![(key, Some(aux.encode()))],
+			fork_choice: ForkChoiceStrategy::Custom(aux.total_difficulty > best_aux.total_difficulty),
+		};
+
+		Ok((import_block, None))
+	}
+}
+
+/// Register the PoW inherent data provider, if not registered already.
+fn register_pow_inherent_data_provider(
+	inherent_data_providers: &InherentDataProviders,
+) -> Result<(), consensus_common::Error> {
+	if !inherent_data_providers.has_provider(&srml_timestamp::INHERENT_IDENTIFIER) {
+		inherent_data_providers
+			.register_provider(srml_timestamp::InherentDataProvider)
+			.map_err(Into::into)
+			.map_err(consensus_common::Error::InherentData)
+	} else {
+		Ok(())
+	}
+}
+
+/// The PoW import queue type.
+pub type PowImportQueue<B> = BasicQueue<B>;
+
+/// Import queue for PoW engine.
+pub fn import_queue<B, C, Algorithm>(
+	block_import: BoxBlockImport<B>,
+	client: Arc<C>,
+	algorithm: Algorithm,
+	inherent_data_providers: InherentDataProviders,
+) -> Result<PowImportQueue<B>, consensus_common::Error> where
+	B: BlockT<Hash=H256>,
+	C: ProvideRuntimeApi + HeaderBackend<B> + BlockOf + ProvideCache<B> + AuxStore,
+	C: Send + Sync + AuxStore + 'static,
+	C::Api: BlockBuilderApi<B>,
+	Algorithm: PowAlgorithm<B> + Send + Sync + 'static,
+{
+	register_pow_inherent_data_provider(&inherent_data_providers)?;
+
+	let verifier = PowVerifier {
+		client: client.clone(),
+		algorithm,
+		inherent_data_providers,
+	};
+
+	Ok(BasicQueue::new(
+		verifier,
+		block_import,
+		None,
+		None
+	))
+}
+
+/// Start the background mining thread for PoW. Note that because PoW mining
+/// is CPU-intensive, it is not possible to use async future to define this.
+/// However, it's not recommended to use background thread in the rest of the
```

Comment:
```suggestion
/// However, it's not recommended to use background threads in the rest of the
```

That said, most PoW mining is offloaded to dedicated hardware, either GPUs or ASICs.  Substrate will interact with such miners using asynchronous RPC.